InternalDocs/stackrefs.md

```diff
@@ -0,0 +1,79 @@
+  - `Py_INT_TAG` - tagged small integer stored directly in the stackref (no heap allocation).
+- Special constants: `PyStackRef_NULL`, `PyStackRef_ERROR`, and embedded `None`/`True`/`False`.
+
+In GIL builds, most objects carry their refcount; tagged borrowed refs skip decref on close. In free
+threading builds, the tag is also used to mark deferred refcounted objects so the GC can see them and
+to avoid refcount contention for short-lived stack values.
+
+## Converting to and from PyObject*
+
+Three conversions control ownership:
+
+- `PyStackRef_FromPyObjectNew(obj)` - create a new reference (INCREF if mortal).
+- `PyStackRef_FromPyObjectSteal(obj)` - take over ownership without changing the count unless the
+  object is immortal.
+- `PyStackRef_FromPyObjectBorrow(obj)` - create a borrowed stackref (never decref on close).
+
+The `obj` argument must not be `NULL`.
+
+Going back to `PyObject*` mirrors this:
+
+- `PyStackRef_AsPyObjectBorrow(ref)` - borrow the underlying pointer
+- `PyStackRef_AsPyObjectSteal(ref)` - transfer ownership from the stackref
+- `PyStackRef_AsPyObjectNew(ref)` - create a new owning reference
+
+Only `PyStackRef_AsPyObjectBorrow` allows ref to be `PyStackRef_NULL`.
+
+## Operations on stackrefs
+
+The interpreter treats `_PyStackRef` as the unit of stack storage. Ownership must be managed with
+the stackref primitives:
+
+- `PyStackRef_DUP` - like `Py_NewRef` for stackrefs; preserves the original.
+- `PyStackRef_Borrow` - create a borrowed stackref from another stackref.
+- `PyStackRef_CLOSE` / `PyStackRef_XCLOSE` - like `Py_DECREF`; invalidates the stackref.
+- `PyStackRef_CLEAR` - like `Py_CLEAR`; closes and sets the stackref to `PyStackRef_NULL`
+- `PyStackRef_MakeHeapSafe` - converts borrowed reference to owning reference
+
+Borrow tracking (for debug builds with `Py_STACKREF_DEBUG`) records who you borrowed from and reports
+double-close, leaked borrows, or use-after-close via fatal errors.
+
+## Borrow-friendly opcodes
+
+The interpreter can push borrowed references directly. For example, `LOAD_FAST_BORROW` loads a local
+variable as a borrowed `_PyStackRef`, avoiding both INCREF and DECREF for the temporary lifetime on
+the evaluation stack.
+
+## Tagged integers on the stack
+
+Small ints can be stored inline with `Py_INT_TAG`, so no heap object is involved. Helpers like
+`PyStackRef_TagInt`, `PyStackRef_UntagInt`, and `PyStackRef_IncrementTaggedIntNoOverflow` operate on
+these values. Type checks use `PyStackRef_IsTaggedInt` and `PyStackRef_LongCheck`.
+
+## Free threading considerations
+
+With `Py_GIL_DISABLED`, `Py_TAG_DEFERRED` is an alias for `Py_TAG_REFCNT`.
+Objects that support deferred reference counting can be pushed to the evaluation
+stack and stored in local variables without directly incrementing the reference
+count because they are only freed during cyclic garbage collection. This avoids
+reference count contention on short-lived values such as methods and types. The GC
```

Comment:
This is confusing to me because methods and types tend to be long-lived.
```suggestion
reference count contention on commonly shared objects such as methods and types. The GC
```